### PR TITLE
feat: add reverse mapping for pool key, store tL, tU

### PR DESCRIPTION
**Pull request by snreynolds**
motivation was to have better interface standardization for ERC721s
(functions dont require PositionConfig to be passed in)

- & you get an onchain lookup of poolKey, tickLower, tickUpper
- & position descriptor can read from config

I personally was against this for gas reasons (mint is 40-50K more expensive) but that should go down with mints on the same pool, and overhead for other ops is around 4-5K.

### CODE REVIEW

File: .forge-snapshots/PositionManager_mint_native.snap [hunk 1]
```diff
@@ -1 +1 @@
-341353
\ No newline at end of file
+370766
```
Comment (saucepoint) [inline .forge-snapshots/PositionManager_mint_native.snap:1]: i regret to inform yall that we have, unfortunately, lost sara to ultrasound money gang 😭 

pouring one out for all my mainnet retail LPs 🙏

Comment (snreynolds) [reply .forge-snapshots/PositionManager_mint_native.snap:1]: 🪦

**Comment by hensha256** [comment]
>(mint is 40-50K more expensive

so its like 8k more for mints on the same pool right? which is the "normal" case?

File: src/libraries/PositionConfig.sol [hunk 2]
```diff
@@ -9,27 +9,3 @@ struct PositionConfig {
     int24 tickLower;
     int24 tickUpper;
 }
-
```
Comment (hensha256) [inline src/libraries/PositionConfig.sol:12]: now that this is just a struct shall it just go into IPositionManager?

File: src/interfaces/ISubscriber.sol [hunk 3]
```diff
@@ -7,23 +7,15 @@ import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
...
     /// @param tokenId the token ID of the position
-    /// @param config details about the position
     /// @param data additional data passed in by the caller
-    function notifySubscribe(uint256 tokenId, PositionConfig memory config, bytes memory data) external;
+    function notifySubscribe(uint256 tokenId, bytes memory data) external;
     /// @param tokenId the token ID of the position
-    /// @param config details about the position
     /// @param data additional data passed in by the caller
-    function notifyUnsubscribe(uint256 tokenId, PositionConfig memory config, bytes memory data) external;
+    function notifyUnsubscribe(uint256 tokenId, bytes memory data) external;
```
Comment (hensha256) [inline src/interfaces/ISubscriber.sol:14]: i feel like the subscriber would still want this information? imo it should still be sent

File: src/PositionManager.sol [hunk 4]
```diff
@@ -110,23 +110,32 @@ contract PositionManager is
...
+    // TODO: Move to custom type, packed memory.
+    struct PositionInfo {
+        // lower 25 bytes of the poolId
+        bytes25 poolId;
+        int24 tickLower;
+        int24 tickUpper;
+        bool hasSubscriber;
+    }
+
+    mapping(uint256 tokenId => PositionInfo info) internal positionInfo;
```
Comment (hensha256) [inline src/PositionManager.sol:133]: public?

Comment (hensha256) [reply src/PositionManager.sol:133]: unless im missing something? how will position descriptor call to fetch this information?

File: src/PositionManager.sol [hunk 5]
```diff
@@ -110,23 +110,32 @@ contract PositionManager is
...
+    mapping(bytes32 poolId => PoolKey poolKey) internal poolKeys;
```
Comment (hensha256) [inline src/PositionManager.sol:134]: bytes25?

Comment (hensha256) [reply src/PositionManager.sol:134]: public?

File: src/PositionManager.sol [hunk 6]
```diff
@@ -432,8 +430,7 @@ contract PositionManager is
         liquidity = poolManager.getPositionLiquidity(config.poolKey.toId(), positionId);
     }
 
-    /// @inheritdoc IPositionManager
-    function getPositionConfigId(uint256 tokenId) external view returns (bytes32) {
-        return positionConfigs[tokenId].getConfigId();
+    function _toConfig(PositionInfo memory info) internal view returns (PositionConfig memory config) {
```
Comment (hensha256) [inline src/PositionManager.sol:433]: should PositionInfo by a library with this function in it?
im open

File: src/interfaces/ISubscriber.sol (hunk 3 above)
Comment (snreynolds) [reply src/interfaces/ISubscriber.sol:14]: subscriber can just read it tho so if they dont need it its worse. i think we should just remove it and now its consistent with notifyTransfer

File: src/PositionManager.sol (hunk 6 above)
Comment (snreynolds) [reply src/PositionManager.sol:433]: yeah was gonna move some things around just wanted to check people liked the basic idea

**Comment by snreynolds** [comment]
> > (mint is 40-50K more expensive
> 
> so its like 8k more for mints on the same pool right? which is the "normal" case?

not sure I can add a test but sounds about right it definitely wont be 40-50K

File: src/interfaces/ISubscriber.sol (hunk 3 above)
Comment (snreynolds) [reply src/interfaces/ISubscriber.sol:14]: also it means more standardized to ERC721s not just PositionManager which is I think the point of this whole change no?

File: src/PositionManager.sol [hunk 7]
```diff
@@ -259,37 +236,56 @@ contract PositionManager is
...
     function _increase(
         uint256 tokenId,
-        PositionConfig calldata config,
         uint256 liquidity,
         uint128 amount0Max,
         uint128 amount1Max,
         bytes calldata hookData
-    ) internal onlyIfApproved(msgSender(), tokenId) onlyValidConfig(tokenId, config) {
+    ) internal onlyIfApproved(msgSender(), tokenId) {
+        (PositionInfo memory info, PoolKey memory poolKey) = getPoolPositionInfo(tokenId);
```
Comment (saucepoint) [inline src/PositionManager.sol:244]: are there gas savings if we dont load the entirety of PositionInfo? we're not using the truncated poolId obviously, and the ticks are just repacked to IPoolManager.ModifyLiquidityParams

something like this could be quite cool
```solidity
(
    PoolKey memory key,
    IPoolManager.ModifyLiquidityParams memory liquidityParams,
    bool hasSubscriber
) = veryAwesomeFunction(tokenId);

liquidityParams.liquidityDelta = liquidityChange;
```

theres also ways to pack it further since PoolKey is 528 bits, so we can fit tL/tU/hasSubscriber in the remaining bits of the 3rd word

**Review by saucepoint — COMMENTED** [review]
despite the massive diff, i quite like this

something about not checking subscriber on mint feels nice, same with maybe making burn only call unsubscribe

File: src/PositionManager.sol [hunk 8]
```diff
@@ -304,37 +300,50 @@ contract PositionManager is
...
 
-        delete positionConfigs[tokenId];
+        delete positionInfo[tokenId];
         // Burn the token.
         _burn(tokenId);
+
+        // TODO: This will be changed from audit issue..
+        if (info.hasSubscriber) {
+            _notifyModifyLiquidity(uint256(tokenId), -(liquidity.toInt256()), feesAccrued);
+        }
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/PositionManager.sol:346]: I guess if burn does call modifyLiquidity then we should

File: src/PositionManager.sol [hunk 9]
```diff
@@ -415,25 +402,47 @@ contract PositionManager is
...
+        poolKey = poolKeys[info.poolId];
     }
 
     /// @inheritdoc IPositionManager
-    function getPositionConfigId(uint256 tokenId) external view returns (bytes32) {
-        return positionConfigs[tokenId].getConfigId();
+    function getPositionLiquidity(uint256 tokenId) public view returns (uint128 liquidity) {
+        (PositionInfo memory info, PoolKey memory poolKey) = getPoolPositionInfo(tokenId);
+        liquidity = _getLiquidity(tokenId, poolKey, info.tickLower, info.tickUpper);
+    }
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/PositionManager.sol:438]: external is a good call

File: test/erc721Permit/ERC721Permit.permit.t.sol [hunk 10]
```diff
@@ -229,7 +229,8 @@ contract ERC721PermitTest is Test {
         assertEq(erc721Permit.nonces(alice, wordPos) & (1 << bitPos), 0);
     }
 
-    function test_fuzz_erc721Permit_SignatureDeadlineExpired(address spender) public {
+    function test_fuzz_erc721Permit_SignatureDeadlineExpired() public {
+        address spender = 0xf30b182eA406CC5A71b8A44e2FE50cC80383F914;
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply test/erc721Permit/ERC721Permit.permit.t.sol:233]: oh i was trying something out.. via-ir breaks this test. even on main

File: src/PositionManager.sol (hunk 7 above)
Comment (snreynolds) [reply src/PositionManager.sol:244]: yeah was thinking the same thing.. I think I'm actually happy with it now that I have a version where the PositionInfo is packed (uint256).. since its the minimum we can basically return

File: src/libraries/PositionConfig.sol [hunk 11]
```diff
@@ -3,33 +3,9 @@ pragma solidity ^0.8.24;
 
 import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
 
-// A PositionConfig is the input for creating and modifying a Position in core, whose truncated hash is set per tokenId
+// A PositionConfig is the input for creating and modifying a Position in core
```
Comment (snreynolds) [inline src/libraries/PositionConfig.sol:6]: Keeping this for now to reduce merge conflicts

later can move to test utils

File: src/interfaces/ISubscriber.sol (hunk 3 above)
Comment (hensha256) [reply src/interfaces/ISubscriber.sol:14]: yeah thats all fair, agreed

File: src/base/Notifier.sol [hunk 12]
```diff
@@ -30,27 +29,23 @@ abstract contract Notifier is INotifier {
...
     modifier onlyIfApproved(address caller, uint256 tokenId) virtual;
-    modifier onlyValidConfig(uint256 tokenId, PositionConfig calldata config) virtual;
-
-    function _positionConfigs(uint256 tokenId) internal view virtual returns (PositionConfigId storage);
+    function _setUnsubscribe(uint256 tokenId) internal virtual;
+    function _setSubscribe(uint256 tokenId) internal virtual;
 
     /// @inheritdoc INotifier
-    function subscribe(uint256 tokenId, PositionConfig calldata config, address newSubscriber, bytes calldata data)
+    function subscribe(uint256 tokenId, address newSubscriber, bytes calldata data)
```
Comment (hensha256) [inline src/base/Notifier.sol:36]: i found the flow of setting subscription/unsubscription very hard to follow

so we have 2 ways of showing that someone is subscribed (as previously), both in the new PositionInfo and in the mapping

The mapping `subscriber` is easy to follow
- look it up in the mapping
- check theyre not subscribed already
- set the new subscriber

To check that position info boolean is subscribed
- external function `subscribe`
- calls virtual function `_subscribe`
- implementation of `_subscribe` is in position manager
- that implementation calls another function `setSubscribe`
- when you read `setSubscribe` it says `/// @ dev this does not actually set any storage`... so its still not set that its subscribed
- then back to position manager to find the actual line that then sets the return data into storage

Comment (hensha256) [reply src/base/Notifier.sol:36]: basically that second flow feels quite unnatural to follow currently, trying to think through how it could be better designed

File: src/PositionManager.sol [hunk 13]
```diff
@@ -110,24 +110,20 @@ contract PositionManager is
...
-    using PositionConfigLibrary for PositionConfig;
     using StateLibrary for IPoolManager;
     using TransientStateLibrary for IPoolManager;
     using SafeCast for uint256;
     using SafeCast for int256;
     using CalldataDecoder for bytes;
     using SlippageCheckLibrary for BalanceDelta;
-    using PositionConfigIdLibrary for PositionConfigId;
+    using PoolKeyChecker for PoolKey;
+    using PositionInfoLibrary for PackedPositionInfo;
```
Comment (hensha256) [inline src/PositionManager.sol:120]: nit: would rather the names were the same like
`PositionInfo` and `PositionInfoLibrary`
or `PackedPositionInfo` and `PackedPositionInfoLibrary`

File: src/PositionManager.sol [hunk 14]
```diff
@@ -110,24 +110,20 @@ contract PositionManager is
...
     /// @dev The ID of the next token that will be minted. Skips 0
     uint256 public nextTokenId = 1;
 
-    mapping(uint256 tokenId => PositionConfigId configId) internal positionConfigs;
-
-    /// @notice an internal getter for PositionConfigId to be used by Notifier
-    function _positionConfigs(uint256 tokenId) internal view override returns (PositionConfigId storage) {
-        return positionConfigs[tokenId];
-    }
+    mapping(uint256 tokenId => PackedPositionInfo info) public positionInfo;
```
Comment (hensha256) [inline src/PositionManager.sol:125]: now that its packed maybe worth having this internal and exposing a getter that returns the elements individually?

File: src/PositionManager.sol [hunk 15]
```diff
@@ -259,37 +232,41 @@ contract PositionManager is
...
     function _increase(
         uint256 tokenId,
-        PositionConfig calldata config,
         uint256 liquidity,
         uint128 amount0Max,
         uint128 amount1Max,
         bytes calldata hookData
-    ) internal onlyIfApproved(msgSender(), tokenId) onlyValidConfig(tokenId, config) {
+    ) internal onlyIfApproved(msgSender(), tokenId) {
+        (PackedPositionInfo info, PoolKey memory poolKey) = getPoolPositionInfo(tokenId);
```
Comment (hensha256) [inline src/PositionManager.sol:240]: maybe im wrong but i imagine it would be decently cheaper to just get these directly from the mappings inline instead of calling a public function

```
PositionInfo info = positionInfo[tokenId];
...
_modifyLiquidity(info, poolKeys[info.poolId()], liquidity.toInt256(), bytes32(tokenId), hookData);
```

File: src/libraries/PoolKeyChecker.sol [hunk 16]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+
+library PoolKeyChecker {
+    // On UniswapV4, the minimum tick spacing is 1.
+    // This means that if the tick spacing is 0, the pool key has not been set.
+    function isEmpty(PoolKey memory poolKey) internal pure returns (bool) {
```
Comment (hensha256) [inline src/libraries/PoolKeyChecker.sol:9]: this library is just used for 1 single function call, and its 1 line of code.
I think I'd personally rather have no library and just do it inline in PosM like

```
        // Store the poolKey if it is not already stored.
        if (poolKeys[info.poolId()].tickSpacing == 0) {
            poolKeys[info.poolId()] = poolKey;
        }
```

File: src/PositionManager.sol [hunk 17]
```diff
@@ -304,35 +281,39 @@ contract PositionManager is
...
-            _modifyLiquidity(config, liquidity.toInt256(), bytes32(tokenId), hookData);
+            _modifyLiquidity(info, poolKey, liquidity.toInt256(), bytes32(tokenId), hookData);
         // Slippage checks should be done on the principal liquidityDelta which is the liquidityDelta - feesAccrued
         (liquidityDelta - feesAccrued).validateMaxIn(amount0Max, amount1Max);
-        positionConfigs[tokenId].setConfigId(config.toId());
 
-        emit MintPosition(tokenId, config);
+        // Store the poolKey if it is not already stored.
+        if (poolKeys[info.poolId()].isEmpty()) {
+            poolKeys[info.poolId()] = poolKey;
```
Comment (hensha256) [inline src/PositionManager.sol:295]: double calculating `info.poolId()`

File: src/PositionManager.sol (hunk 13 above)
Comment (hensha256) [reply src/PositionManager.sol:120]: I prefer the first one i think. Packing it is just an implementation detail - so `PositionInfo` feels better

File: src/PositionManager.sol (hunk 15 above)
Comment (hensha256) [reply src/PositionManager.sol:240]: i do see its neater though, not against keeping it
I'd maybe call it `getPoolAndPositionInfo`

File: src/PositionManager.sol [hunk 18]
```diff
@@ -383,57 +364,78 @@ contract PositionManager is
         if (balance > 0) currency.transfer(to, balance);
     }
 
+    // implementation of abstract function DeltaResolver._pay
+    function _pay(Currency currency, address payer, uint256 amount) internal override {
+        if (payer == address(this)) {
+            // TODO: currency is guaranteed to not be eth so the native check in transfer is not optimal.
```
Comment (hensha256) [inline src/PositionManager.sol:370]: remove the TODO... unless you actually plan to make this change by tuesday?

File: src/PositionManager.sol [hunk 19]
```diff
@@ -304,35 +281,39 @@ contract PositionManager is
...
         if (liquidity > 0) {
             (BalanceDelta liquidityDelta, BalanceDelta feesAccrued) =
-                _modifyLiquidity(config, -(liquidity.toInt256()), bytes32(tokenId), hookData);
+                _modifyLiquidity(info, poolKey, -(liquidity.toInt256()), bytes32(tokenId), hookData);
             // Slippage checks should be done on the principal liquidityDelta which is the liquidityDelta - feesAccrued
             (liquidityDelta - feesAccrued).validateMinOut(amount0Min, amount1Min);
         }
 
-        delete positionConfigs[tokenId];
+        positionInfo[tokenId] = PositionInfoLibrary.EMPTY_POSITION_INFO;
```
Comment (hensha256) [inline src/PositionManager.sol:316]: cant you just do `delete positionInfo[tokenId]`?
and then `EMPTY_POSITION_INFO` is unused

File: src/libraries/PositionInfoLibrary.sol [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+ * A flag to know if the tokenId is subscribed to an address
+ * uint8 hasSubscriber;
+ *
+ * The tickUpper of the position
+ * int24 tickUpper;
+ *
+ * The tickLower of the position
+ * int24 tickLower;
+ *
+ * The truncated poolId
```
Comment (hensha256) [inline src/libraries/PositionInfoLibrary.sol:25]: more description ideally, which end is truncated?

File: src/base/Notifier.sol (hunk 12 above)
Comment (snreynolds) [reply src/base/Notifier.sol:36]: we used to be able to set it in notifiers but bc the mapping is now a uint256 I cant pass a storage flag so it has to be set in manager.

if you have suggestions lmk!

File: src/PositionManager.sol (hunk 17 above)
Comment (snreynolds) [reply src/PositionManager.sol:295]: Yeah I just didnt wanna save a variable cause I was hitting stack too deep like every 5 mins last night

File: src/PositionManager.sol (hunk 18 above)
Comment (snreynolds) [reply src/PositionManager.sol:370]: that todo is on main?

File: src/PositionManager.sol (hunk 19 above)
Comment (snreynolds) [reply src/PositionManager.sol:316]: you can't do that cause its a uint256

File: src/PositionManager.sol (hunk 13 above)
Comment (snreynolds) [reply src/PositionManager.sol:120]: sgtm!

File: src/PositionManager.sol (hunk 14 above)
Comment (snreynolds) [reply src/PositionManager.sol:125]: hm I dont feel strongly about it. For offchain, a simple sdk helper can dissect it. and onchain someone can just import the PositionInfoLibrary?

But if you prefer I can add

File: src/PositionManager.sol (hunk 15 above)
Comment (snreynolds) [reply src/PositionManager.sol:240]: either way. I just did it cause it was repeated in 4 funcs

File: src/libraries/PoolKeyChecker.sol (hunk 16 above)
Comment (snreynolds) [reply src/libraries/PoolKeyChecker.sol:9]: I mostly liked it as a lib cause its not obvious that tickSpacing should be checked so i was just getting it out of the main codepath but i can change if you would prefer

File: src/PositionManager.sol (hunk 18 above)
Comment (hensha256) [reply src/PositionManager.sol:365]: yeah i know, i just think we should remove it given we're finalising periphery

File: src/base/Notifier.sol [hunk 21]
```diff
@@ -30,26 +28,24 @@ abstract contract Notifier is INotifier {
...
         external
         payable
         onlyIfApproved(msg.sender, tokenId)
-        onlyValidConfig(tokenId, config)
     {
-        // will revert below if the user already has a subcriber
-        _positionConfigs(tokenId).setSubscribe();
+        _setSubscribe(tokenId);
```
Comment (hensha256) [inline src/base/Notifier.sol:42]: i think to officially follow checks-effects-interactions as our audits have brought up this should go after the check that
```
if (_subscriber != NO_SUBSCRIBER)
```

so it does that check, then sets subscribers in both ways, and then does the call

File: src/base/Notifier.sol [hunk 22]
```diff
@@ -59,34 +55,24 @@ abstract contract Notifier is INotifier {
...
-    function unsubscribe(uint256 tokenId, PositionConfig calldata config, bytes calldata data)
-        external
-        payable
-        onlyIfApproved(msg.sender, tokenId)
-        onlyValidConfig(tokenId, config)
-    {
-        _positionConfigs(tokenId).setUnsubscribe();
+    function unsubscribe(uint256 tokenId, bytes calldata data) external payable onlyIfApproved(msg.sender, tokenId) {
+        _setUnsubscribe(tokenId);
         ISubscriber _subscriber = subscriber[tokenId];
```
Comment (hensha256) [inline src/base/Notifier.sol:60]: should this revert if they dont already have a subscriber?

File: src/libraries/PositionInfoLibrary.sol [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    function setSubscribe(PositionInfo info) internal pure returns (PositionInfo _info) {
+        assembly ("memory-safe") {
+            _info := or(info, SET_SUBSCRIBE)
+        }
+    }
+
+    /// @dev this does not actually set any storage
+    function setUnsubscribe(PositionInfo info) internal pure returns (PositionInfo _info) {
+        assembly ("memory-safe") {
+            _info := and(info, not(SET_SUBSCRIBE))
```
Comment (hensha256) [inline src/libraries/PositionInfoLibrary.sol:79]: doesnt this set it to `0xfffffffe`? wouldnt we rather just set it to 0?

File: src/interfaces/IPositionManager.sol [hunk 24]
```diff
@@ -29,16 +27,12 @@ interface IPositionManager is INotifier {
...
     /// @return liquidity the position's liquidity, as a liquidityAmount
     /// @dev this value can be processed as an amount0 and amount1 by using the LiquidityAmounts library
-    function getPositionLiquidity(uint256 tokenId, PositionConfig calldata config)
-        external
-        view
-        returns (uint128 liquidity);
+    function getPositionLiquidity(uint256 tokenId) external view returns (uint128 liquidity);
+
+    /// @param tokenId the ERC721 tokenId
+    /// @return  a uint256 packed value holding information about the position including the range (tickLower, tickUpper)
```
Comment (hensha256) [inline src/interfaces/IPositionManager.sol:35]:
```suggestion
    /// @return a uint256 packed value holding information about the position including the range (tickLower, tickUpper)
```

File: src/PositionManager.sol (hunk 18 above)
Comment (hensha256) [reply src/PositionManager.sol:365]: oh i see its literally from this file - how come youve moved it up? the diff confused me

Comment (hensha256) [reply src/PositionManager.sol:365]: esp if we're getting this audited would be good to have a minimal diff

File: src/PositionManager.sol (hunk 19 above)
Comment (snreynolds) [reply src/PositionManager.sol:313]: sorry maybe its actually cause its a custom type.. im not sure but I tried this and it doesnt work

File: test/libraries/CalldataDecoder.t.sol [hunk 25]
```diff
@@ -19,48 +19,39 @@ contract CalldataDecoderTest is Test {
...
             decoder.decodeBurnParams(params);
 
         assertEq(tokenId, _tokenId);
         assertEq(hookData, _hookData);
-        _assertEq(_config, config);
         assertEq(amount0Min, _amount0Min);
         assertEq(amount1Min, _amount1Min);
     }
 
+    // TODO: fix stack too deep here
```
Comment (hensha256) [inline test/libraries/CalldataDecoder.t.sol:54]: is it fixed now?

File: src/libraries/PositionConfig.sol (hunk 11 above)
Comment (hensha256) [reply src/libraries/PositionConfig.sol:6]: woudl rather delete now that we arent using it

File: test/libraries/CalldataDecoder.t.sol (hunk 25 above)
Comment (snreynolds) [reply test/libraries/CalldataDecoder.t.sol]: yes i just removed that comment! thanks

File: test/mocks/MockCalldataDecoder.sol [hunk 26]
```diff
@@ -1,40 +1,38 @@
...
+    struct MintParams {
+        PoolKey poolKey;
+        int24 tickLower;
+        int24 tickUpper;
+        uint256 liquidity;
+        uint128 amount0Max;
+        uint128 amount1Max;
+        address owner;
+        bytes hookData;
+    }
```
Comment (saucepoint) [inline test/mocks/MockCalldataDecoder.sol:22]: this is struct-ified to avoid IR? if so, maybe worth adding a small comment why its struct-ified in tests and only for mint

File: src/PositionManager.sol [hunk 27]
```diff
@@ -412,25 +395,40 @@ contract PositionManager is
...
 
+    /// @notice an internal helper used by Notifier
+    function _setSubscribe(uint256 tokenId) internal override {
+        positionInfo[tokenId] = positionInfo[tokenId].setSubscribe();
+    }
+
+    /// @notice an internal helper used by Notifier
+    function _setUnsubscribe(uint256 tokenId) internal override {
+        positionInfo[tokenId] = positionInfo[tokenId].setUnsubscribe();
+    }
```
Comment (saucepoint) [inline src/PositionManager.sol:406]: mini nit: should we rename this to `_setSubscribed` and `_setUnsubscribed` just to be clearer. Feel like subscribe/subscriber is quite overloaded

File: src/libraries/PositionInfoLibrary.sol [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+    uint256 internal constant SET_UNSUBSCRIBE = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00;
+    uint256 internal constant SET_SUBSCRIBE = 0x01;
+    uint8 internal constant TICK_LOWER_OFFSET = 8;
+    uint8 internal constant TICK_UPPER_OFFSET = 32;
+
+    function poolId(PositionInfo info) internal pure returns (bytes25 _poolId) {
+        assembly ("memory-safe") {
+            _poolId := and(MASK_UPPER_200_BITS, info)
+        }
+    }
```
Comment (saucepoint) [inline src/libraries/PositionInfoLibrary.sol:50]: Lets add a natspec saying `info.poolId()` is NOT compatible with `PoolId` on core.

i'm tempted to suggest calling it something else like `truncatedPoolId()` but 🤮

**Review by saucepoint — APPROVED** [review]
very clean work per usual 😌

File: src/libraries/PositionInfoLibrary.sol (hunk 28 above)
Comment (snreynolds) [reply src/libraries/PositionInfoLibrary.sol:50]: Yeah good call out, I'll add natspec 😛

File: test/mocks/MockCalldataDecoder.sol (hunk 26 above)
Comment (snreynolds) [reply test/mocks/MockCalldataDecoder.sol:22]: yes will do

Comment (snreynolds) [reply test/mocks/MockCalldataDecoder.sol:22]: and yes to avoid ir lol

File: src/PositionManager.sol [hunk 29]
```diff
@@ -412,25 +395,40 @@ contract PositionManager is
...
     /// @inheritdoc IPositionManager
-    function getPositionLiquidity(uint256 tokenId, PositionConfig calldata config)
-        public
-        view
-        returns (uint128 liquidity)
-    {
-        bytes32 positionId =
-            Position.calculatePositionKey(address(this), config.tickLower, config.tickUpper, bytes32(tokenId));
-        liquidity = poolManager.getPositionLiquidity(config.poolKey.toId(), positionId);
+    function getPoolAndPositionInfo(uint256 tokenId) public view returns (PositionInfo info, PoolKey memory poolKey) {
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/PositionManager.sol:415]: yep!

**Comment by saucepoint** [comment]
reminder: consider SB-86, hasSubscriber should revert if the token does not exist

stance: probably ugly to enforce, and adds gas. BUT should be natspec'd in hasSubscriber function

File: src/PositionManager.sol [hunk 30]
```diff
@@ -301,35 +276,41 @@ contract PositionManager is
...
         (liquidityDelta - feesAccrued).validateMaxIn(amount0Max, amount1Max);
-        positionConfigs[tokenId].setConfigId(config.toId());
 
-        emit MintPosition(tokenId, config);
+        bytes25 poolId = info.poolId();
+        // Store the poolKey if it is not already stored.
+        // On UniswapV4, the minimum tick spacing is 1, which means that if the tick spacing is 0, the pool key has not been set.
+        if (poolKeys[poolId].tickSpacing == 0) {
+            poolKeys[poolId] = poolKey;
+        }
```
Comment (saucepoint) [inline src/PositionManager.sol:293]: let's do this write prior to `_modifyLiquidity` for CEI anyway (the hook still has access to tick range and poolKey regardless of CEI)

File: src/PositionManager.sol [hunk 31]
```diff
@@ -301,35 +276,41 @@ contract PositionManager is
...
-                _modifyLiquidity(config, -(liquidity.toInt256()), bytes32(tokenId), hookData);
+                _modifyLiquidity(info, poolKey, -(liquidity.toInt256()), bytes32(tokenId), hookData);
             // Slippage checks should be done on the principal liquidityDelta which is the liquidityDelta - feesAccrued
             (liquidityDelta - feesAccrued).validateMinOut(amount0Min, amount1Min);
         }
 
-        delete positionConfigs[tokenId];
+        positionInfo[tokenId] = PositionInfoLibrary.EMPTY_POSITION_INFO;
         // Burn the token.
         _burn(tokenId);
```
Comment (saucepoint) [inline src/PositionManager.sol:315]: thinking this logic should run prior to `_modifyLiquidity` for CEI

**Review by saucepoint — COMMENTED** [review]
some CEI violations in _mint / _burn

https://github.com/spearbit-audits/review-uniswap-v4/issues/93

File: src/PositionManager.sol (hunk 31 above)
Comment (saucepoint) [reply src/PositionManager.sol:315]: solmate _burn might be save to leave it, but positionInfo may want to happen before?

File: src/libraries/PositionInfoLibrary.sol [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+    function initialize(PoolKey memory _poolKey, int24 _tickLower, int24 _tickUpper)
+        internal
+        pure
+        returns (PositionInfo info)
+    {
+        bytes25 _poolId = bytes25(PoolId.unwrap(_poolKey.toId()));
+        assembly {
+            info :=
+                or(
+                    or(and(MASK_UPPER_200_BITS, _poolId), shl(TICK_UPPER_OFFSET, and(MASK_24_BITS, _tickUpper))),
```
Comment (0age) [inline src/libraries/PositionInfoLibrary.sol:100]: I think this mask is redundant as it's already bytes25

File: src/PositionManager.sol (hunk 30 above)
Comment (snreynolds) [reply src/PositionManager.sol:302]: ah yeah that makes sense

File: test/mocks/MockCalldataDecoder.sol [hunk 33]
```diff
@@ -1,40 +1,39 @@
...
 import {CalldataDecoder} from "../../src/libraries/CalldataDecoder.sol";
 import {IV4Router} from "../../src/interfaces/IV4Router.sol";
 import {Currency} from "@uniswap/v4-core/src/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
 
 // we need to use a mock contract to make the calls happen in calldata not memory
 contract MockCalldataDecoder {
     using CalldataDecoder for bytes;
 
+    // This is used to avoid compiling with via-ir.
```
Comment (saucepoint) [inline test/mocks/MockCalldataDecoder.sol:13]: image

File: test/libraries/PositionInfoLibrary.t.sol [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    function test_fuzz_initialize_setSubscribed(PoolKey memory poolKey, int24 tickLower, int24 tickUpper) public pure {
+        PositionInfo info = PositionInfoLibrary.initialize(poolKey, tickLower, tickUpper);
+        assertEq(info.hasSubscriber(), false);
+        info = info.setSubscribe();
+        assertEq(info.hasSubscriber(), true);
+        assertEq(info.tickLower(), tickLower);
+        assertEq(info.tickUpper(), tickUpper);
+        assertEq(info.poolId(), bytes25(PoolId.unwrap(poolKey.toId())));
```
Comment (saucepoint) [inline test/libraries/PositionInfoLibrary.t.sol:31]: library itself looks good, but as an insane person we should assert that the tickLower/tickUpper/poolId is NOT changed by `.setSubscribe`

i.e. asserting that the info data is the same before/after `.setSubscribe`

File: src/PositionManager.sol [hunk 35]
```diff
@@ -310,39 +285,46 @@ contract PositionManager is
...
+        bytes25 poolId = info.poolId();
+        if (poolKeys[poolId].tickSpacing == 0) {
+            poolKeys[poolId] = poolKey;
+        }
 
-        emit MintPosition(tokenId, config);
+        // fee delta can be ignored as this is a new position
+        (BalanceDelta liquidityDelta,) =
+            _modifyLiquidity(info, poolKey, liquidity.toInt256(), bytes32(tokenId), hookData);
+        (liquidityDelta).validateMaxIn(amount0Max, amount1Max);
```
Comment (hensha256) [inline src/PositionManager.sol:302]:
```suggestion
        liquidityDelta.validateMaxIn(amount0Max, amount1Max);
```

File: src/PositionManager.sol [hunk 36]
```diff
@@ -391,24 +373,26 @@ contract PositionManager is
...
+                tickUpper: info.tickUpper(),
                 liquidityDelta: liquidityChange,
                 salt: salt
             }),
             hookData
         );
 
-        if (positionConfigs[uint256(salt)].hasSubscriber()) {
-            _notifyModifyLiquidity(uint256(salt), config, liquidityChange, feesAccrued);
+        // TODO: Audit issue for burn, decide if we want to keep this and also unsubscribe.
```
Comment (hensha256) [inline src/PositionManager.sol:393]: remove

**Review by hensha256 — DISMISSED** [review]

**Review by hensha256 — APPROVED** [review]

**Review by hensha256 — APPROVED** [review]
